### PR TITLE
add pagination

### LINKED ISSUE #242: Enable pagination for the get_assets endpoint
The [get_assets](https://github.com/canonical/assets.ubuntu.com/blob/98d338cc1268097ad0f86501c78e8f9f4d968378/webapp/views.py#L154) endpoint doesn't seem to support pagination. This can lead to very long wait times and unnecessary server load if a search query is sufficiently generic. 

It would be nice if this endpoint supported pagination (i.e.., accept `start` and `limit` parameters, or perhaps just a `page` parameter with a predefined number of results per page) to only fetch a subset of results matching a query, rather than the entire resultset.

Then the front-end could be updated to pull in pages of data rather than the entire resultset. Perhaps the [Vanilla Pagination pattern](https://vanillaframework.io/docs/patterns/pagination) would be of good use here.

### PR DESCRIPTION
**Pull request by goulinkh**
## Done

- Add pagination with per_page option between 20 (default) and 100
- Parse available file extensions from the list of files in the DB

## QA

- Run the app locally, this needs a database and files server so dotrun doesn't work ATM:
```bash
# start the db and files server locally
docker compose up -d
# make sure you are using python3.8 (latest version supported ATM)
# If you know of a way to change python version please do, otherwise feel free
# to follow this guide: https://mise.jdx.dev/getting-started.html
# once installed run this in the assets project: mise use python@3.8

# create a virtual environment
python3 -m venv .venv

# install requirements
pip install -r requirements

# Install dotenv tool
pip install python-dotenv

# install frontend dependencies
yarn install


# Finally, run the app
dotenv -f .env.local -f .env run yarn start
```
- Create a few assets, (or contact me and I will share a DB file with you to import a list of assets)
### UI QA
- Make a search with one character for example: ` ` (space)
- Make sure that the pagination is working as expected
- Make sure that the file types list is working
### API QA
- First, you need to have an API token to be able to make requests
```bash
# list the existing tokens and choose one
dotenv -f .env.local -f .env run flask token list
# if the list is empty, run this command to create one
dotenv -f .env.local -f .env run flask token create test
```
- Make sure the this shows at most 20 items by default
```bash
curl http://0.0.0.0:8017/v1\?token\={your-token}
```
-  Make sure that this returns 100 items per page:
```bash
curl http://0.0.0.0:8017/v1\?token\={your-token}\&per_page\=100
```
- Make sure that this returns 20 items per page (more that max 100):
```bash
curl http://0.0.0.0:8017/v1\?token\={your-token}\&per_page\=101
```
- Make sure that the `page` option works:
```bash
curl http://0.0.0.0:8017/v1\?token\={your-token}\&per_page\=100\&page=2
```

## Issue / Card

Fixes #242

## Screenshots

image

### CODE REVIEW

**Comment by immortalcodes** [comment]
image
I suppose there used to be support for several file types

**Comment by immortalcodes** [comment]
Also there is already work going on assets improvement and we have some features on this branch
https://github.com/canonical/assets.ubuntu.com/tree/feature_assets_improvement
Pagination was something that we were planning to do in this cycle and get it to this branch.

**Comment by goulinkh** [comment]
> image I suppose there used to be support for several file types

The new implementation gets the extensions from the database, having an empty list means that there are no assets in the database.

**Comment by immortalcodes** [comment]
image
Slight inconsistency in numbering

**Comment by immortalcodes** [comment]
Sorting with filename will not be useful if it is using hash codes and not the actual file names.
image

File: webapp/services.py [hunk 1]
```diff
@@ -36,9 +47,16 @@ def find_assets(self, file_type="%", tag=None, query=None):
...
             tag_condition = Asset.tags.any(Tag.name == tag)
             conditions.append(tag_condition)
-        assets = db_session.query(Asset).filter(*conditions).yield_per(100)
-        for asset in assets:
-            yield asset
+        assets = (
+            db_session.query(Asset)
+            .filter(*conditions)
+            .order_by(order_by.desc() if desc_order else order_by)
+            .offset((max(page, 1) - 1) * max(per_page, 10))
```
Comment (immortalcodes) [inline webapp/services.py:54]: Why a maximum here, isn't per_page always defined to be 20 when called, even so why a max between 10 when by default we call it with 20

**Review by immortalcodes — CHANGES_REQUESTED** [review]
This is really good work and is going to be quite helpful for entire team.
My comments are mostly edge cases and nitpicking. Once those seem fine we can merge it.
Thanks for this Awesome PR!

**Review by immortalcodes — APPROVED** [review]

**Comment by immortalcodes** [comment]
LGTM!
These features will be loved!

**Pull request merged**